### PR TITLE
[ipcamera] Fix Hikvision cameras stay offline when a 401 reply is given with no www-authenticate header

### LINKED ISSUE #15413: [ipcamera] CONFIGURATION_ERROR Camera gave no WWW-Authenticate: Your login details must be wrong.
Somehow, after a while (sometimes around 24h or more) my ipcamera things go offline with:

```
CONFIGURATION_ERROR
Camera gave no WWW-Authenticate: Your login details must be wrong.
```

I have NOT changed anything on my cameras (i.e. no firmware update or so). The issue started to appear with an openHAB snapshot from around may 2023. I tried to pinpoint it to a commit, but there are no real changes in the binding around that time.

It could be due to an update of the netty library and a different handling of requests in it.

I added some more logs where the things goes offline with that error and this is what response the binding got when it ran into the error:

```
2023-08-10 16:15:24.047 [WARN ] [ipcamera.internal.MyNettyAuthHandler] - Stefan: Response from camera: DefaultHttpResponse(decodeResult: failure(java.lang.IllegalArgumentException: Validation failed for header 'WWW-Authenticate'), version: HTTP/1.1)
HTTP/1.1 401 Unauthorized
Date: Thu, 10 Aug 2023 16:15:23 GMT
Server: webserver
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block
Content-Length: 178
Content-Type: text/html
Connection: close
2023-08-10 16:15:24.047 [WARN ] [ipcamera.internal.MyNettyAuthHandler] - Stefan: do NOT processAuth() with '' on url='/ISAPI/Streaming/channels/101/picture' via method='GET'
2023-08-10 16:15:24.047 [WARN ] [ipcamera.internal.MyNettyAuthHandler] - Stefan: Got header: 'Date' with value 'Thu, 10 Aug 2023 16:15:23 GMT'
2023-08-10 16:15:24.047 [WARN ] [ipcamera.internal.MyNettyAuthHandler] - Stefan: Got header: 'Server' with value 'webserver'
2023-08-10 16:15:24.048 [WARN ] [ipcamera.internal.MyNettyAuthHandler] - Stefan: Got header: 'X-Content-Type-Options' with value 'nosniff'
2023-08-10 16:15:24.048 [WARN ] [ipcamera.internal.MyNettyAuthHandler] - Stefan: Got header: 'X-Frame-Options' with value 'SAMEORIGIN'
2023-08-10 16:15:24.048 [WARN ] [ipcamera.internal.MyNettyAuthHandler] - Stefan: Got header: 'X-XSS-Protection' with value '1; mode=block'
2023-08-10 16:15:24.048 [WARN ] [ipcamera.internal.MyNettyAuthHandler] - Stefan: Got header: 'Content-Length' with value '178'
2023-08-10 16:15:24.048 [WARN ] [ipcamera.internal.MyNettyAuthHandler] - Stefan: Got header: 'Content-Type' with value 'text/html'
2023-08-10 16:15:24.048 [WARN ] [ipcamera.internal.MyNettyAuthHandler] - Stefan: Got header: 'Connection' with value 'close'
```

It happens here in the code:
https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/MyNettyAuthHandler.java#L163

So `authenticate` is empty and as i am printing out all headers, we can see that there is no header containing `WWW-Authenticate`.

The `processAuth` looks like it could handle an empty `authenticate` parameter, but as the logic explicitly deals with that case and runs into the error instead, I wonder why that is ;)

If I deactivate and reactivate my thing, it comes back online and works for hours/days, before it runs into the issue again. So I am wondering whether a retry should be implemented if a camera (falsely?) responds with 401 out of a sudden to some request...

### PR DESCRIPTION
**Pull request by t2000**
It might happen that a IP camera (Hikvision) replies with HTTP 401 but does not add a WWW-Authenticate header for our reconnect logic.

This fix reconnects the ThingHandler to the camera in such cases.

Fixes #15413

### CODE REVIEW

**Comment by lolodomo** [comment]
@Skinah : I let you check if that is fine for you.

**Review by kaikreuzer — APPROVED** [review]
lgtm, especially as cameras often behave buggy and thus retrying is the best option.

**Comment by jlaur** [comment]
FYI, there are some recent discussions/investigations in the linked issue.

**Comment by kaikreuzer** [comment]
Ah, thanks @jlaur, I missed that discussion.

**Comment by t2000** [comment]
I have updated the MR according to the discussion here: https://github.com/openhab/openhab-addons/issues/15413#issuecomment-1848441460

File: bundles/org.openhab.binding.ipcamera/src/main/resources/OH-INF/thing/thing-types.xml
Comment (lolodomo) [inline bundles/org.openhab.binding.ipcamera/src/main/resources/OH-INF/thing/thing-types.xml:1989]: Do not forget to adjust README to describe this new parameter and to add the two new entries (label and description) in the ipcamera.properties file.

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.ipcamera/src/main/resources/OH-INF/thing/thing-types.xml
Comment (t2000) [reply bundles/org.openhab.binding.ipcamera/src/main/resources/OH-INF/thing/thing-types.xml:1989]: Good point, thanks.
I have added a new commit to address that.

File: bundles/org.openhab.binding.ipcamera/src/main/resources/OH-INF/i18n/ipcamera.properties
Comment (lolodomo) [inline bundles/org.openhab.binding.ipcamera/src/main/resources/OH-INF/i18n/ipcamera.properties:386]: Please synchronize label and description here with those in XML description.

**Review by lolodomo — CHANGES_REQUESTED** [review]

**Comment by lolodomo** [comment]
@Skinah : please have a look.

File: bundles/org.openhab.binding.ipcamera/src/main/resources/OH-INF/i18n/ipcamera.properties
Comment (lolodomo) [reply bundles/org.openhab.binding.ipcamera/src/main/resources/OH-INF/i18n/ipcamera.properties:386]: Simply run: mvn i18n:generate-default-tramslations

Comment (t2000) [reply bundles/org.openhab.binding.ipcamera/src/main/resources/OH-INF/i18n/ipcamera.properties]: Thanks for the hint: Done.

**Review by lolodomo — APPROVED** [review]
LGTM, thank you

**Comment by t2000** [comment]
> In the discussion of the issue thread, I did state that this can be automated by asking the camera if it supports inputs and then removing the channels if the camera replies that it does not. 

Automatic behavior is nice and I am a fan of it if it works ;-)
In theory we can detect if a camera supports external inputs, but from the community thread we already saw that cameras might respond that they support external I/Os while in reality they don't. For such cases we still would need a configuration to override the automatic behavior :(

> This is what I would like to discuss before this is merged, is there a reason why another config is added that will cause a breaking change?

It is certainly not breaking because of the new config parameter, as there is a default for that in case it is not set.

However, I see the point that the default should probably be set to `true`, so everything stays as is, so for those users who are not having trouble, nothing changes. And if you do NOT have external alarm inputs you have to explicitly configure this new config parameter to `false`.

Would that be a solution (you could add the automatic detection later anyhow if you like and keep this override for faulty cameras)?

> please do not feel disheartened by this comment as I appreciate your work
I absolutely don't feel like that. I am always in favor for the "right" solution. :-)

**Comment by Skinah** [comment]
Make a call to `/ISAPI/System/IO/capabilities`

If it is zero like this one or it is missing, you know the camera has no IO outputs.

```
<IOOutputPortNums>0</IOOutputPortNums>
```
See
https://community.openhab.org/t/ipcamera-binding-doing-erroneous-isapi-request-to-hikvision-camera/147345/3

It is confusing as they have also posted the output of the `status` api call.

**Comment by t2000** [comment]
Reverted to reconnect, see discussion.

**Comment by lolodomo** [comment]
Waiting for @Skinah approval to merge, we are still in time for 4.1...

**Comment by Skinah** [comment]
LGTM

**Review by lolodomo — APPROVED** [review]
Thank you

**Pull request merged**
